pkg/northbound/gnmi/set.go

```diff
@@ -17,23 +17,119 @@ package gnmi
 import (
 	"context"
 	"fmt"
-	"log"
-	"strings"
-	"time"
-
 	"github.com/docker/docker/pkg/namesgenerator"
 	"github.com/onosproject/onos-config/pkg/manager"
 	"github.com/onosproject/onos-config/pkg/store"
 	"github.com/onosproject/onos-config/pkg/store/change"
 	"github.com/onosproject/onos-config/pkg/utils"
 	"github.com/openconfig/gnmi/proto/gnmi"
 	"github.com/openconfig/gnmi/proto/gnmi_ext"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"log"
+	"strings"
+	"time"
 )
 
 type mapTargetUpdates map[string]map[string]string
 type mapTargetRemoves map[string][]string
 type mapNetworkChanges map[store.ConfigName]change.ID
 
+// Set implements gNMI Set
+func (s *Server) Set(ctx context.Context, req *gnmi.SetRequest) (*gnmi.SetResponse, error) {
+	// There is only one set of extensions in Set request, regardless of number of
+	// updates
+	var (
+		netcfgchangename string // May be specified as 100 in extension
+		version          string // May be specified as 101 in extension
+		deviceType       string // May be specified as 102 in extension
+	)
+
+	targetUpdates := make(mapTargetUpdates)
+	targetRemoves := make(mapTargetRemoves)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	//Update
+	for _, u := range req.GetUpdate() {
+		target := u.Path.GetTarget()
+		targetUpdates[target] = s.doUpdateOrReplace(u, targetUpdates)
+	}
+
+	//Replace
+	for _, u := range req.GetReplace() {
+		target := u.Path.GetTarget()
+		targetUpdates[target] = s.doUpdateOrReplace(u, targetUpdates)
+	}
+
+	//Delete
+	for _, u := range req.GetDelete() {
+		target := u.GetTarget()
+		targetRemoves[target] = s.doDelete(u, targetRemoves)
+	}
+
+	for _, ext := range req.GetExtension() {
+		if ext.GetRegisteredExt().GetId() == GnmiExtensionNetwkChangeID {
+			netcfgchangename = string(ext.GetRegisteredExt().GetMsg())
+		} else if ext.GetRegisteredExt().GetId() == GnmiExtensionVersion {
+			version = string(ext.GetRegisteredExt().GetMsg())
+		} else if ext.GetRegisteredExt().GetId() == GnmiExtensionDeviceType {
+			deviceType = string(ext.GetRegisteredExt().GetMsg())
+		} else {
+			return nil, fmt.Errorf("Unexpected extension %d = '%s' in Set()",
```

Comment:
Is there a way to set a code on this - I don't think it would be Internal though - more like InvalidArgument

Reply:
done